src/server/pps/server/api_server.go

```diff
@@ -1519,19 +1448,197 @@ func (a *apiServer) getLogsFromStats(pachClient *client.APIClient, request *pps.
+	// shutdown for pod log goros
+	if err != io.EOF {
+		return err
+	}
+	return nil
+}
+
+func (a *apiServer) getLogsForRC(pachClient *client.APIClient, request *pps.GetLogsRequest, containerName, rcName string, logCh chan *pps.LogMessage) error {
+	defer close(logCh)
+
+	pods, err := a.rcPodsFromLogsRequest(pachClient, request, rcName)
+	if err != nil {
+		return err
+	}
+
+	// Sort the pods to make sure that the order of log lines is stable
+	sort.Sort(podSlice(pods))
+
+	for _, pod := range pods {
+		if err := a.getLogsFromPod(pachClient.Ctx(), request, pod, containerName, logCh); err != nil && err != io.EOF {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *apiServer) rcPodsFromLogsRequest(pachClient *client.APIClient, request *pps.GetLogsRequest, rcName string) ([]v1.Pod, error) {
+	// If an RC name hasn't been passed in, fetch one now
+	curRCName := rcName
+	if curRCName == "" {
+		var pipelineInfo *pps.PipelineInfo
+		var err error
+
+		if request.Pipeline != nil {
+			pipelineInfo, err = a.inspectPipeline(pachClient, request.Pipeline.Name)
+			if err != nil {
+				return nil, fmt.Errorf("could not get pipeline information for %s: %v", request.Pipeline.Name, err)
+			}
+		} else if request.Job != nil {
+			// If user provides a job, lookup the pipeline from the job info, and then
+			// get the pipeline RC
+			var jobPtr pps.EtcdJobInfo
+			err = a.jobs.ReadOnly(pachClient.Ctx()).Get(request.Job.ID, &jobPtr)
+			if err != nil {
+				return nil, fmt.Errorf("could not get job information for \"%s\": %v", request.Job.ID, err)
+			}
+			pipelineInfo, err = a.inspectPipeline(pachClient, jobPtr.Pipeline.Name)
+			if err != nil {
+				return nil, fmt.Errorf("could not get pipeline information for %s: %v", jobPtr.Pipeline.Name, err)
+			}
+		}
+
+		curRCName = ppsutil.PipelineRcName(pipelineInfo.Pipeline.Name, pipelineInfo.Version)
+	}
+
+	// Get pods managed by the RC we're scraping (either pipeline or pachd)
+	pods, err := a.rcPods(curRCName)
+	if err != nil {
+		return nil, fmt.Errorf("could not get pods in rc \"%s\" containing logs: %s", curRCName, err.Error())
+	}
+	if len(pods) == 0 {
+		return nil, fmt.Errorf("no pods belonging to the rc \"%s\" were found", curRCName)
+	}
+
+	return pods, nil
+}
+
+func (a *apiServer) getLogsFromPod(ctx context.Context, request *pps.GetLogsRequest, pod v1.Pod, containerName string, logCh chan *pps.LogMessage) error {
+	tailLines := &request.Tail
+	if *tailLines <= 0 {
+		tailLines = nil
+	}
+
+	// Get full set of logs from the pod
+	stream, err := a.env.GetKubeClient().CoreV1().Pods(a.namespace).GetLogs(
+		pod.ObjectMeta.Name, &v1.PodLogOptions{
+			Container: containerName,
+			Follow:    request.Follow,
+			TailLines: tailLines,
+		}).Timeout(10 * time.Second).Stream()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		// Close the stream if the context is done. This is needed because we
+		// otherwise risk blocking indefinitely on `scanner.Scan()`.
+		<-ctx.Done()
```

Comment:
I *think* we have to manually close `ctx` in the success case if we want to avoid leaking this goro. Just checking if that happens? (I couldn't figure out where right away)

Reply:
This context comes from an error group. In the success case:

1) This function will return `io.EOF`.
2) The error group will receive the error from one of its workers, and close `ctx`.
3) All of the other workers will receive the `ctx` close here, and close their streams.

Reply:
:+1: Thanks! I didn't realize that that was how `errgroup.WithContext()` worked so thanks for the explanation.

Reply:
Yeah, I learned about that from working on this, and there's some other places where it might be useful: I think you had a todo in either the pachd or worker `main.go` to address the default error group behavior of waiting until all goros finish.